packages/services/api/src/modules/audit-logs/providers/audit-log-recorder.ts

```diff
@@ -0,0 +1,66 @@
+import { randomUUID } from 'node:crypto';
+import { Injectable, Scope } from 'graphql-modules';
+import { captureException } from '@sentry/node';
+import { sql as c_sql, ClickHouse } from '../../operations/providers/clickhouse-client';
+import { Logger } from '../../shared/providers/logger';
+import { AuditLogRecordEvent, auditLogSchema, UserContextSchema } from './audit-logs-types';
+
+@Injectable({
+  scope: Scope.Operation,
+  global: true,
+})
+/**
+ * Responsible for recording audit log events and storing them in ClickHouse
+ */
+export class AuditLogRecorder {
+  private logger: Logger;
+
+  constructor(
+    logger: Logger,
+    private clickHouse: ClickHouse,
+  ) {
+    this.logger = logger.child({ source: 'AuditLogRecorder' });
+  }
+
+  async record(data: AuditLogRecordEvent): Promise<void> {
+    try {
+      const { eventType, organizationId, userEmail, userId } = data;
+      this.logger.debug('Creating audit log event', { eventType });
+
+      const parsedMetadata = auditLogSchema.parse(data);
+      const parsedContext = UserContextSchema.parse(data.user);
+      const eventMetadata = JSON.stringify({
+        ...parsedMetadata,
+        ...parsedContext,
+      });
```

Comment:
No need to have `UserContextSchema` here. You can just do the following, we don't need a zod model here for extracting the properties. Also `email` would be redundant since we already have a `"audit_logs"."user_email"` column. For isolation purposes, let's also move the metadata properties to a nested user object.

```
      const parsedContext = UserContextSchema.parse(data.user);
      const eventMetadata = JSON.stringify({
        ...parsedMetadata,
        user: {
          fullName: data.user.fullName,
          displayName: data.user.displayName,
          provider: data.user.provider,
        },
      });
```